src/theory/uf/lambda_lift.cpp

```diff
@@ -0,0 +1,187 @@
+  {
+    TrustNode trn = lift(node);
+    lems.push_back(SkolemLemma(trn, skolem));
+  }
+  // if no proofs, return lemma with no generator
+  if (d_epg == nullptr)
+  {
+    return TrustNode::mkTrustRewrite(node, skolem);
+  }
+  Node eq = node.eqNode(skolem);
+  return d_epg->mkTrustedRewrite(
+      node, skolem, PfRule::MACRO_SR_PRED_INTRO, {eq});
+}
+
+Node LambdaLift::getLambdaFor(TNode skolem) const
+{
+  NodeNodeMap::const_iterator it = d_lambdaMap.find(skolem);
+  if (it == d_lambdaMap.end())
+  {
+    return Node::null();
+  }
+  return it->second;
+}
+
+Node LambdaLift::getAssertionFor(TNode node)
+{
+  TNode skolem = getSkolemFor(node);
+  if (skolem.isNull())
+  {
+    return Node::null();
+  }
+  Kind k = node.getKind();
+  Node assertion;
+  if (k == LAMBDA)
+  {
+    NodeManager* nm = NodeManager::currentNM();
+    // The new assertion
+    std::vector<Node> children;
+    // bound variable list
+    children.push_back(node[0]);
+    // body
+    std::vector<Node> skolem_app_c;
+    skolem_app_c.push_back(skolem);
+    skolem_app_c.insert(skolem_app_c.end(), node[0].begin(), node[0].end());
+    Node skolem_app = nm->mkNode(APPLY_UF, skolem_app_c);
+    children.push_back(skolem_app.eqNode(node[1]));
+    // axiom defining skolem
+    assertion = nm->mkNode(FORALL, children);
+
+    // Lambda lifting is trivial to justify, hence we don't set a proof
+    // generator here. In particular, replacing the skolem introduced
+    // here with its original lambda ensures the new assertion rewrites
+    // to true.
+    // For example, if (lambda y. t[y]) has skolem k, then this lemma is:
+    //   forall x. k(x)=t[x]
+    // whose witness form rewrites
+    //   forall x. (lambda y. t[y])(x)=t[x] --> forall x. t[x]=t[x] --> true
+  }
+  return assertion;
+}
+
+Node LambdaLift::getSkolemFor(TNode node)
+{
+  Node skolem;
+  Kind k = node.getKind();
+  if (k == LAMBDA)
+  {
+    // if a lambda, do lambda-lifting
+    if (!expr::hasFreeVar(node))
+    {
+      Trace("rtf-proof-debug")
+          << "RemoveTermFormulas::run: make LAMBDA skolem" << std::endl;
+      // Make the skolem to represent the lambda
+      NodeManager* nm = NodeManager::currentNM();
+      SkolemManager* sm = nm->getSkolemManager();
+      skolem = sm->mkPurifySkolem(
+          node,
+          "lambdaF",
+          "a function introduced due to term-level lambda removal");
+    }
+  }
+  return skolem;
+}
+
+TrustNode LambdaLift::betaReduce(TNode node) const
+{
+  Kind k = node.getKind();
+  if (k == APPLY_UF)
+  {
+    Node op = node.getOperator();
+    Node opl = getLambdaFor(op);
+    if (!opl.isNull())
+    {
+      std::vector<Node> args(node.begin(), node.end());
+      Node app = betaReduce(opl, args);
+      Trace("uf-lazy-ll") << "Beta reduce: " << node << " -> " << app
+                          << std::endl;
+      return TrustNode::mkTrustRewrite(node, app, nullptr);
+    }
+  }
+  // otherwise, unchanged
+  return TrustNode::null();
```

Comment:
Did you mean to return `node` here or `null`? I guess it depends on how you call this function. Better to document the expected behavior in the header file.